test: Invalid test results files when testing is terminated.

I have a Jenkins job that is executing the mpich tests.  When the Jenkins job hits its time limit and kills the mpich testing jobs, the summary.xml and summary.junit.xml files are invalid and incomplete.

The summary.xml file begins with :
```
<?xml version='1.0' ?>
<?xml-stylesheet href="TestResults.xsl" type="text/xsl" ?>
<MPITESTRESULTS>
<DATE>2022-08-03-21-03</DATE>
<MPITEST>
<NAME>attrt</NAME>
<ARGS></ARGS>
<NP>2</NP>
<WORKDIR>./attr</WORKDIR>
<STATUS>pass</STATUS>
<TIME>0.325302839279175</TIME>
</MPITEST>
```

and ends with (depending on when it gets terminated):
```
</MPITEST>
<MPITEST>
<NAME>allred3</NAME>
<ARGS></ARGS>
<NP>10</NP>
<WORKDIR>./coll</WORKDIR>

```
The last <MPITEST> entry is not completed, i.e. ends with a </MPITEST>.
And it does not have (even with a successful completion of the tests):
`</MPITESTRESULTS>
`
For the summary.juinit.xml file (in a successful test run) it begins with:
```
<testsuites>
  <testsuite failures="350"
             errors="0"
             skipped="26"
             tests="2464"
             date="2022-07-30-12-45"
             name="summary_junit_xml">
```

and ends with:
```
  </testsuite>
</testsuites>
```

However, if the testing is aborted, i.e. terminated before it completes successfully, these lines are not in the file.

I tested using Jenkins' "Run with timeout" job step, and the console log are -
```
...
Build timed out (after 45 minutes). Marking the build as failed.
Build was aborted
Archiving artifacts
Unexpected output in mt_improbe_sendrecv_huge: 
Unexpected output in mt_improbe_sendrecv_huge: ===================================================================================
Unexpected output in mt_improbe_sendrecv_huge: =   BAD TERMINATION OF ONE OF YOUR APPLICATION PROCESSES
Unexpected output in mt_improbe_sendrecv_huge: =   PID 28362 RUNNING AT pmrs-gpu-240-02.cels.anl.gov
Unexpected output in mt_improbe_sendrecv_huge: =   EXIT CODE: 15
Unexpected output in mt_improbe_sendrecv_huge: =   CLEANING UP REMAINING PROCESSES
Unexpected output in mt_improbe_sendrecv_huge: =   YOU CAN IGNORE THE BELOW CLEANUP MESSAGES
Unexpected output in mt_improbe_sendrecv_huge: ===================================================================================
Unexpected output in mt_improbe_sendrecv_huge: YOUR APPLICATION TERMINATED WITH THE EXIT STRING: Terminated (signal 15)
Unexpected output in mt_improbe_sendrecv_huge: This typically refers to a problem with your application.
Unexpected output in mt_improbe_sendrecv_huge: Please see the FAQ page for debugging suggestions
Program mt_improbe_sendrecv_huge exited without No Errors
Killed: 
make: *** [testing] Error 255
Recording test results
...
```
What timeout  mechanism do you use? Do you see `Killed:` in the console log? That is something printed by `runtests` from the signal handler.

--- a/test/mpi/runtests
+++ b/test/mpi/runtests
@@ -1218,8 +1218,10 @@ sub RunTestPassed {
         print TAPOUT "ok $g_total_seen - $workdir/$programname ${np} # time=$runtime\n";
     }
     if ($junitoutput) {
+        my $classname = $workdir;
+        $classname =~ s/^\.\///;
         # Use %05d since the Junit plugin in Jenkins sort tests by name
-        printf JUNITOUT "    <testcase name=\"%05d - $workdir/$programname ${np} ${progArgs} ${progEnv}\" time=\"$runtime\"></testcase>\n", $g_total_seen;
+        printf JUNITOUT "    <testcase classname=\"$classname\" name=\"%05d - $workdir/$programname ${np} ${progArgs} ${progEnv}\" time=\"$runtime\"></testcase>\n", $g_total_seen;
     }
 }
 sub RunTestFailed {
@@ -1302,7 +1304,9 @@ sub RunTestFailed {
             $xfailstr = " # TODO $test_opt->{xfail}";
             $testtag  = "skipped";
         }
-        printf JUNITOUT "    <testcase name=\"%05d - $workdir/$programname ${np} ${progArgs} ${progEnv}\" time=\"$runtime\">\n", $g_total_seen;
+        my $classname = $workdir;
+        $classname =~ s/^\.\///;
+        printf JUNITOUT "    <testcase classname=\"$classname\" name=\"%05d - $workdir/$programname ${np} ${progArgs} ${progEnv}\" time=\"$runtime\">\n", $g_total_seen;
         print JUNITOUT "      <${testtag} type=\"TestFailed\"\n";
         print JUNITOUT "               message=\"not ok $g_total_seen - $workdir/$programname ${np}${xfailstr}\"><![CDATA[";
         print JUNITOUT "not ok $g_total_seen - $workdir/$programname ${np}${xfailstr}\n";
@@ -1346,7 +1350,9 @@ sub SkippedTest {
         print TAPOUT "ok $g_total_seen - $workdir/$programname $np  # SKIP $reason\n";
     }
     if ($junitoutput) {
-        printf JUNITOUT "    <testcase name=\"%05d - $workdir/$programname ${np} ${progArgs} ${progEnv}\">\n", $g_total_seen;
+        my $classname = $workdir;
+        $classname =~ s/^\.\///;
+        printf JUNITOUT "    <testcase classname=\"$classname\" name=\"%05d - $workdir/$programname ${np} ${progArgs} ${progEnv}\">\n", $g_total_seen;
         print JUNITOUT "      <skipped type=\"TodoTestSkipped\">\n";
         print JUNITOUT "             message=\"$reason\"><![CDATA[ok $g_total_seen - $workdir/$programname $np  # SKIP $reason]]></skipped>\n";
         print JUNITOUT "    </testcase>\n";

--- a/test/mpi/runtests
+++ b/test/mpi/runtests
@@ -1218,7 +1218,8 @@ sub RunTestPassed {
         print TAPOUT "ok $g_total_seen - $workdir/$programname ${np} # time=$runtime\n";
     }
     if ($junitoutput) {
-        print JUNITOUT "    <testcase name=\"$g_total_run - $workdir/$programname ${np} ${progArgs} ${progEnv}\" time=\"$runtime\"></testcase>\n";
+        # Use %05d since the Junit plugin in Jenkins sort tests by name
+        printf JUNITOUT "    <testcase name=\"%05d - $workdir/$programname ${np} ${progArgs} ${progEnv}\" time=\"$runtime\"></testcase>\n", $g_total_seen;
     }
 }
 sub RunTestFailed {
@@ -1301,10 +1302,10 @@ sub RunTestFailed {
             $xfailstr = " # TODO $test_opt->{xfail}";
             $testtag  = "skipped";
         }
-        print JUNITOUT "    <testcase name=\"$g_total_run - $workdir/$programname ${np} ${progArgs} ${progEnv}\" time=\"$runtime\">\n";
+        printf JUNITOUT "    <testcase name=\"%05d - $workdir/$programname ${np} ${progArgs} ${progEnv}\" time=\"$runtime\">\n", $g_total_seen;
         print JUNITOUT "      <${testtag} type=\"TestFailed\"\n";
-        print JUNITOUT "               message=\"not ok $g_total_run - $workdir/$programname ${np}${xfailstr}\"><![CDATA[";
-        print JUNITOUT "not ok $g_total_run - $workdir/$programname ${np}${xfailstr}\n";
+        print JUNITOUT "               message=\"not ok $g_total_seen - $workdir/$programname ${np}${xfailstr}\"><![CDATA[";
+        print JUNITOUT "not ok $g_total_seen - $workdir/$programname ${np}${xfailstr}\n";
         print JUNITOUT "  ---\n";
         print JUNITOUT "  Directory: $workdir\n";
         print JUNITOUT "  File: $programname\n";
@@ -1345,7 +1346,7 @@ sub SkippedTest {
         print TAPOUT "ok $g_total_seen - $workdir/$programname $np  # SKIP $reason\n";
     }
     if ($junitoutput) {
-        print JUNITOUT "    <testcase name=\"$g_total_seen - $workdir/$programname ${np} ${progArgs} ${progEnv}\">\n";
+        printf JUNITOUT "    <testcase name=\"%05d - $workdir/$programname ${np} ${progArgs} ${progEnv}\">\n", $g_total_seen;
         print JUNITOUT "      <skipped type=\"TodoTestSkipped\">\n";
         print JUNITOUT "             message=\"$reason\"><![CDATA[ok $g_total_seen - $workdir/$programname $np  # SKIP $reason]]></skipped>\n";
         print JUNITOUT "    </testcase>\n";

--- a/test/mpi/runtests
+++ b/test/mpi/runtests
@@ -1097,6 +1097,13 @@ sub TestErrFatal {
 #  RunTestFailed, RunTestPassed - Call after test
 #  RunPostMsg               - Call at end of each test
 #
+sub on_SIGINT {
+    chdir $g_cwd or die;
+    RunPostMsg();
+    CloseOutputs();
+    die "Killed: $!\n";
+}
+
 sub OpenOutputs {
     if ($xmloutput) {
         open( XMLOUT, ">$g_opt{xmlfile}" ) || die "Cannot open $g_opt{xmlfile}\n";
@@ -1122,6 +1129,10 @@ sub OpenOutputs {
     if ($junitoutput) {
         open( JUNITOUT, ">$g_opt{junitfile}" ) || die "Cannot open $g_opt{junitfile}\n";
     }
+
+    # Catch signals to wrap the output
+    $SIG{INT} = \&on_SIGINT;
+    $SIG{TERM} = \&on_SIGINT;
 }
 
 sub CloseOutputs {
@@ -1144,7 +1155,7 @@ sub CloseOutputs {
 
         # the second pass: insert the header
         # Note: the field "errors" is not used now, but reserved for future uses.
-        open my $JUNITIN,  '<',  $g_opt{junitfile}      or die "Can't read old file: $!";
+        open my $JUNITIN,  '<',  $g_opt{junitfile}      or die "Can't read old file $g_opt{junitfile}: $!";
         open my $JUNITOUTNEW, '>', "$g_opt{junitfile}.new" or die "Can't write new file: $!";
         my $date = `date "+%Y-%m-%dT%H:%M:%S"`;
         $date =~ s/\r?\n//;

--- a/test/mpi/runtests
+++ b/test/mpi/runtests
@@ -115,6 +115,7 @@ $g_opt{junitfile} = '';
 my $xmloutput;
 my $tapoutput;
 my $junitoutput;
+my ($xml_state, $tap_state, $junit_state);
 
 # Total number of tests checked and run
 my $g_total_seen = 0;         # $g_ok_count + $g_err_count + $g_skip_count
@@ -239,15 +240,6 @@ foreach $_ (@ARGV) {
     elsif (/--?xmlfile=(.*)/) {
         $g_opt{xmlfile}   = $1;
         $xmloutput = 1;
-        open( XMLOUT, ">$g_opt{xmlfile}" ) || die "Cannot open $g_opt{xmlfile}\n";
-        my $date = `date "+%Y-%m-%d-%H-%M"`;
-        $date =~ s/\r?\n//;
-        # MPISOURCE can be used to describe the source of MPI for this
-        # test.
-        print XMLOUT "<?xml version='1.0' ?>$g_opt{newline}";
-        print XMLOUT "<?xml-stylesheet href=\"TestResults.xsl\" type=\"text/xsl\" ?>$g_opt{newline}";
-        print XMLOUT "<MPITESTRESULTS>$g_opt{newline}";
-        print XMLOUT "<DATE>$date</DATE>$g_opt{newline}";
     }
     elsif (/--?noxmlclose/) {
         $g_opt{noxmlclose} = 1;
@@ -255,19 +247,10 @@ foreach $_ (@ARGV) {
     elsif (/--?tapfile=(.*)/) {
         $g_opt{tapfile} = $1;
         $tapoutput = 1;
-        open( TAPOUT, ">$g_opt{tapfile}" ) || die "Cannot open $g_opt{tapfile}\n";
-        my $date = `date "+%Y-%m-%d-%H-%M"`;
-        $date =~ s/\r?\n//;
-        print TAPOUT "TAP version 13\n";
-        print TAPOUT "# MPICH test suite results (TAP format)\n";
-        print TAPOUT "# date ${date}\n";
-        # we do not know at this point how many tests will be run, so do
-        # not print a test plan line like "1..450" until the very end
     }
     elsif (/--?junitfile=(.*)/) {
         $g_opt{junitfile} = $1;
         $junitoutput = 1;
-        open( JUNITOUT, ">$g_opt{junitfile}" ) || die "Cannot open $g_opt{junitfile}\n";
     }
     else {
         print STDERR "Unrecognized argument $_\n";
@@ -297,6 +280,8 @@ if ($g_opt{batchRun}) {
     open( BATOUT, ">$g_opt{batrundir}/runtests.batch" ) || die "Could not open $g_opt{batrundir}/runtests.batch\n";
 }
 
+OpenOutputs();
+
 #
 # Process any files
 
@@ -323,45 +308,7 @@ else {
 }
 
 RunTests(\@all_tests);
-
-if ($xmloutput && !$g_opt{noxmlclose}) { 
-    print XMLOUT "</MPITESTRESULTS>$g_opt{newline}";
-    close XMLOUT; 
-}
-
-if ($tapoutput) {
-    print TAPOUT "1..$g_total_seen\n";
-    close TAPOUT;
-}
-
-if ($junitoutput) {
-    print JUNITOUT "    <system-out></system-out>\n";
-    print JUNITOUT "    <system-err></system-err>\n";
-    print JUNITOUT "  </testsuite>\n";
-    print JUNITOUT "</testsuites>\n";
-    close JUNITOUT;
-
-    # the second pass: insert the header
-    # Note: the field "errors" is not used now, but reserved for future uses.
-    open my $JUNITIN,  '<',  $g_opt{junitfile}      or die "Can't read old file: $!";
-    open my $JUNITOUTNEW, '>', "$g_opt{junitfile}.new" or die "Can't write new file: $!";
-    my $date = `date "+%Y-%m-%dT%H:%M:%S"`;
-    $date =~ s/\r?\n//;
-    print $JUNITOUTNEW "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n";
-    print $JUNITOUTNEW "<testsuites>\n";
-    print $JUNITOUTNEW "  <testsuite failures=\"$g_err_count\"\n";
-    print $JUNITOUTNEW "             errors=\"0\"\n";
-    print $JUNITOUTNEW "             skipped=\"$g_skip_count\"\n";
-    print $JUNITOUTNEW "             tests=\"$g_total_run\"\n";
-    print $JUNITOUTNEW "             timestamp=\"${date}\"\n";
-    print $JUNITOUTNEW "             name=\"summary_junit_xml\">\n";
-    while( <$JUNITIN> ) {
-        print $JUNITOUTNEW $_;
-    }
-    close $JUNITIN;
-    close $JUNITOUTNEW;
-    move("$g_opt{junitfile}.new","$g_opt{junitfile}");
-}
+CloseOutputs();
 
 # Output a summary:
 if ($g_opt{batchRun}) {
@@ -1145,10 +1092,79 @@ sub TestErrFatal {
 
 # ----------------------------------------------------------------------------
 # Output routines:
+#  OpenOutputs               - Open report files and print initial lines
 #  RunPreMsg( programname, np, workdir ) - Call before running a program
 #  RunTestFailed, RunTestPassed - Call after test
 #  RunPostMsg               - Call at end of each test
 #
+sub OpenOutputs {
+    if ($xmloutput) {
+        open( XMLOUT, ">$g_opt{xmlfile}" ) || die "Cannot open $g_opt{xmlfile}\n";
+        my $date = `date "+%Y-%m-%d-%H-%M"`;
+        $date =~ s/\r?\n//;
+        # MPISOURCE can be used to describe the source of MPI for this
+        # test.
+        print XMLOUT "<?xml version='1.0' ?>$g_opt{newline}";
+        print XMLOUT "<?xml-stylesheet href=\"TestResults.xsl\" type=\"text/xsl\" ?>$g_opt{newline}";
+        print XMLOUT "<MPITESTRESULTS>$g_opt{newline}";
+        print XMLOUT "<DATE>$date</DATE>$g_opt{newline}";
+    }
+    if ($tapoutput) {
+        open( TAPOUT, ">$g_opt{tapfile}" ) || die "Cannot open $g_opt{tapfile}\n";
+        my $date = `date "+%Y-%m-%d-%H-%M"`;
+        $date =~ s/\r?\n//;
+        print TAPOUT "TAP version 13\n";
+        print TAPOUT "# MPICH test suite results (TAP format)\n";
+        print TAPOUT "# date ${date}\n";
+        # we do not know at this point how many tests will be run, so do
+        # not print a test plan line like "1..450" until the very end
+    }
+    if ($junitoutput) {
+        open( JUNITOUT, ">$g_opt{junitfile}" ) || die "Cannot open $g_opt{junitfile}\n";
+    }
+}
+
+sub CloseOutputs {
+    if ($xmloutput) {
+        if (!$g_opt{noxmlclose}) {
+            print XMLOUT "</MPITESTRESULTS>$g_opt{newline}";
+        }
+        close XMLOUT;
+    }
+    if ($tapoutput) {
+        print TAPOUT "1..$g_total_seen\n";
+        close TAPOUT;
+    }
+    if ($junitoutput) {
+        print JUNITOUT "    <system-out></system-out>\n";
+        print JUNITOUT "    <system-err></system-err>\n";
+        print JUNITOUT "  </testsuite>\n";
+        print JUNITOUT "</testsuites>\n";
+        close JUNITOUT;
+
+        # the second pass: insert the header
+        # Note: the field "errors" is not used now, but reserved for future uses.
+        open my $JUNITIN,  '<',  $g_opt{junitfile}      or die "Can't read old file: $!";
+        open my $JUNITOUTNEW, '>', "$g_opt{junitfile}.new" or die "Can't write new file: $!";
+        my $date = `date "+%Y-%m-%dT%H:%M:%S"`;
+        $date =~ s/\r?\n//;
+        print $JUNITOUTNEW "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n";
+        print $JUNITOUTNEW "<testsuites>\n";
+        print $JUNITOUTNEW "  <testsuite failures=\"$g_err_count\"\n";
+        print $JUNITOUTNEW "             errors=\"0\"\n";
+        print $JUNITOUTNEW "             skipped=\"$g_skip_count\"\n";
+        print $JUNITOUTNEW "             tests=\"$g_total_run\"\n";
+        print $JUNITOUTNEW "             timestamp=\"${date}\"\n";
+        print $JUNITOUTNEW "             name=\"summary_junit_xml\">\n";
+        while( <$JUNITIN> ) {
+            print $JUNITOUTNEW $_;
+        }
+        close $JUNITIN;
+        close $JUNITOUTNEW;
+        move("$g_opt{junitfile}.new","$g_opt{junitfile}");
+    }
+}
+
 sub RunPreMsg {
     my ($test_opt) = @_;
     my $programname = $test_opt->{name};

--- a/test/mpi/runtests
+++ b/test/mpi/runtests
@@ -345,14 +345,15 @@ if ($junitoutput) {
     # Note: the field "errors" is not used now, but reserved for future uses.
     open my $JUNITIN,  '<',  $g_opt{junitfile}      or die "Can't read old file: $!";
     open my $JUNITOUTNEW, '>', "$g_opt{junitfile}.new" or die "Can't write new file: $!";
-    my $date = `date "+%Y-%m-%d-%H-%M"`;
+    my $date = `date "+%Y-%m-%dT%H:%M:%S"`;
     $date =~ s/\r?\n//;
+    print $JUNITOUTNEW "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n";
     print $JUNITOUTNEW "<testsuites>\n";
     print $JUNITOUTNEW "  <testsuite failures=\"$g_err_count\"\n";
     print $JUNITOUTNEW "             errors=\"0\"\n";
     print $JUNITOUTNEW "             skipped=\"$g_skip_count\"\n";
     print $JUNITOUTNEW "             tests=\"$g_total_run\"\n";
-    print $JUNITOUTNEW "             date=\"${date}\"\n";
+    print $JUNITOUTNEW "             timestamp=\"${date}\"\n";
     print $JUNITOUTNEW "             name=\"summary_junit_xml\">\n";
     while( <$JUNITIN> ) {
         print $JUNITOUTNEW $_;

